### DISCUSSION TITLE
Replace page orientation

### DISCUSSION THREAD

**Discussion by dantard** [pymupdf/PyMuPDF#2384] [Looking for help] [answered]
Hello, 
I have a PDF file where the orientation of the pages (that returned by `page.rotation`) is 270 degrees while I can see the page vertically as I need to. 
I'd need to change the orientation of the page to 0 *without* rotating the page, that is, leave the page with the current visible orientation but also have `page.rotation` return 0.

I also tried setting the `/Rotate` xref but it has, as expected, the same result as using `page.set_rotation()`.

Is there any easy way of doing that? 

Thanks in advance.

**Comment by JorjMcKie** [comment, marked as answer]
Yes, there is:
```python
import fitz

doc = fitz.open()  # new output file
src = fitz.open("input.pdf")  # source file

for src_page in src:  # iterate over input pages
    src_rect = src_page.rect  # source page rect
    w, h = src_rect.br  # save its width, height
    src_rot = src_page.rotation  # save source rotation
    src_page.set_rotation(0)  # set rotation to 0 temporarily
    page = doc.new_page(width=w, height=h)  # make output page
    page.show_pdf_page(  # insert source page
        page.rect,
        src,
        src_page.number,
        rotate=-src_rot,  # use reversed original rotation
    )

doc.ez_save("output.pdf")
```

**Reply by sebastianvao** [reply]
I don't know you, I'll probably never be able to see you in my life. But I want you to know that I love you.

Thank you so much

**Reply by negfrequency** [reply]
I want to echo the sentiment of the previous poster. I spent about 5 hours on this problem before finding this solution. thank you!

**Comment by dantard** [comment]
Thank you very much, that's exactly was I was looking for.

**Comment by dantard** [comment]
There would be a way of doing the same keeping annotations, widgets and so on?
I just checked that when using `document.insert_pdf()` (instead of `page.show_pdf_page()`) the page maintains its original orientation in the new document, so it is not useful in this case.

**Reply by JorjMcKie** [reply]
You are right: annotations and widgets of source pages are not part of the new, rotation-free page.
This has a reason:
These object types are **_not part of the page._** That is different from graphics, images, standard text etc.
Therefore, annots and fields can be maintained with a different authorization level.

While there technically also are ways to achieve the effect you want, the effort is considerably higher and would require a real utility.

**Reply by JorjMcKie** [reply]
However: If you do not absolutely need the annotations and fields **_as annotations / fields_**, but only their appearance in form of standard page content, then you can first convert the source PDF using `doc.convert_to_pdf()`.

Given a bit of luck (compatible fonts especially - no Type3 etc.), tthis could be the solution.
